contracts/IAggregator.sol

```diff
@@ -8,9 +8,29 @@ import "./UserOperation.sol";
  */
 interface IAggregator {
 
-  /**
-   * validate aggregated signature.
-   * revert if the aggregated signature does not match the given list of operations.
+    /**
+     * validate aggregated signature.
+     * revert if the aggregated signature does not match the given list of operations.
+     */
+    function validateSignatures(UserOperation[] calldata userOps, bytes calldata signature) external view;
+
+    /**
+     * validate signature of a single userOp
+     * This method is called after EntryPoint.simulateUserOperation() returns an aggregator.
+     * First it validates the signature over the userOp. then it return data to be used when creating the handleOps:
+     * @param userOp the userOperation received from the user.
+   * @return sigForUserOp the value to put into the signature field of the userOp when calling handleOps.
+   *    (usually empty, unless wallet and aggregator support some kind of "multisig"
+   * @return sigForAggregation the value to pass (for all wallets) to aggregateSignatures()
+   */
+    function validateUserOpSignature(UserOperation calldata userOp) external view returns (bytes memory sigForUserOp, bytes memory sigForAggregation);
+
+    /**
+     * aggregate multiple signatures into a single value.
+     * This method is called off-chain to calculate the signature to pass with handleOps()
+     * bundler MAY use optimized custom code perform this aggregation
```

Comment:
Is it legal for an "aggregator" not to implement the aggregation on-chain in solidity and instead **require** that bundlers use native code to perform aggregation? I assume there may be a signature scheme that is not implemented yet or impossible to implement inside the EVM.

Reply:
yes, it MAY leave these unimplemented (or revert). But this requires a bundler that aware of this specific signature scheme and has the required off-chain signature validation/aggregation code.